File: 8/8-a.pl
#!/usr/bin/perl

#
# Copyright (C) 2023 Joelle Maslak
# All Rights Reserved - See License
#

use JTM::Boilerplate 'script';

MAIN: {
    my $steps_str = <<>>;
    chomp($steps_str);
    my (@steps) = split //, $steps_str;
    my $trash = <<>>;

    my %nodes;
    while (my $line = <<>>) {
        chomp($line);
        my ($node, $left, $right) = ($line =~ m/^(...) = \((...), (...)\)$/);
        $nodes{$node} = { left => $left, right => $right };
    }

    my $node = "AAA";
    my $step = 0;
    while ($node ne "ZZZ") {
        $step++;
        my $direction = shift(@steps);
        push @steps, $direction;

        if ($direction eq "L") {
            $node = $nodes{$node}->{left};
        } else {
            $node = $nodes{$node}->{right};
        }
    }

    say("Steps: $step");
}

File: 8/8-a.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
)

type Node struct {
	left  string
	right string
}

func main() {
	reParse := regexp.MustCompile("^(...) = [(](...), (...)[)]$")
	scanner := bufio.NewScanner(os.Stdin)

	scanner.Scan()
	directions := scanner.Text()
	scanner.Scan()

	nodes := make(map[string]Node)
	for scanner.Scan() {
		line := scanner.Text()
		matches := reParse.FindStringSubmatch(line)

		nodes[matches[1]] = Node{left: matches[2], right: matches[3]}
	}

	node := "AAA"
	step := 0
	for node != "ZZZ" {
		dir := directions[step%len(directions)]
		if dir == 'L' {
			node = nodes[node].left
		} else {
			node = nodes[node].right
		}
		step++
	}

	fmt.Print("Steps: ")
	fmt.Println(step)
}
